exporter/azuredataexplorerexporter/adx_exporter.go

```diff
@@ -0,0 +1,278 @@
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	"go.uber.org/zap"
+)
+
+// adxDataProducer uses the ADX client to perform ingestion
+type adxDataProducer struct {
+	client        *kusto.Client       // client for logs , traces and metrics
+	ingestor      ingest.Ingestor     // ingestion for  logs, traces and metrics
+	ingestoptions []ingest.FileOption // Options for the ingestion
+	logger        *zap.Logger         // Loggers for tracing the flow
+}
+
+const nextline = "\n"
+
+const (
+	// Scope name
+	scopename = "scope.name"
+	// Scope version
+	scopeversion = "scope.version"
+)
+
+// given the full metrics , extract each metric , resource attributes and scope attributes. Individual metric mapping is sent on to metricdata mapping
+func (e *adxDataProducer) metricsDataPusher(ctx context.Context, metrics pmetric.Metrics) error {
+	transformedadxmetrics, err := rawMetricsToAdxMetrics(ctx, metrics, e.logger)
+	metricsBuffer := make([]string, len(transformedadxmetrics))
+	if err != nil {
+		e.logger.Error("Error transforming metrics to ADX metric format.", zap.Error(err))
+		return err
+	}
+	// Since the transform succeeded ,  using the option for ingestion ingest the data into ADX
+	for idx, tm := range transformedadxmetrics {
+		adxmetricjsonstring, err := jsoniter.MarshalToString(tm)
+		if err != nil {
+			e.logger.Error("Error performing serialization of data.", zap.Error(err))
+		}
+		metricsBuffer[idx] = adxmetricjsonstring
+	}
+	if len(metricsBuffer) != 0 {
+		if err := e.ingestData(metricsBuffer); err != nil {
+			return err
+		}
+	}
+	metricsflushed := len(transformedadxmetrics)
+	e.logger.Sugar().Infof("Flushing %d metrics to sink", metricsflushed)
+	return nil
+}
+
+func (e *adxDataProducer) ingestData(b []string) error {
+
+	ingestreader := strings.NewReader(strings.Join(b, nextline))
+
+	if _, err := e.ingestor.FromReader(context.Background(), ingestreader, e.ingestoptions...); err != nil {
+		e.logger.Error("Error performing managed data ingestion.", zap.Error(err))
+		return err
+	}
+	return nil
+}
+
+func (e *adxDataProducer) logsDataPusher(ctx context.Context, logData plog.Logs) error {
+	resourcelogs := logData.ResourceLogs()
+	var logsBuffer []string
+	for i := 0; i < resourcelogs.Len(); i++ {
+		resource := resourcelogs.At(i)
+		scopelogs := resourcelogs.At(i).ScopeLogs()
+		for j := 0; j < scopelogs.Len(); j++ {
+			scope := scopelogs.At(j)
+			logs := scopelogs.At(j).LogRecords()
+			for k := 0; k < logs.Len(); k++ {
+				logdata := logs.At(k)
+				transformedadxlog := mapToAdxLog(resource.Resource(), scope.Scope(), logdata, e.logger)
+				adxlogjsonbytes, err := jsoniter.MarshalToString(transformedadxlog)
+				if err != nil {
+					e.logger.Error("Error performing serialization of data.", zap.Error(err))
+				}
+				logsBuffer = append(logsBuffer, adxlogjsonbytes)
+			}
+		}
+	}
+	// Takes care of the residual data.
```

Comment:
I wouldn't call this "residual data", since this is all the data - we don't ingest before.